### PR TITLE
Implement Deepseek-V3 model skeleton

### PR DESCRIPTION
**Pull request by wwwjn**
## Contents
1. Attention module
2. MoE module (note: I only implemented the naive routing, not the "node limit routing" strategy)
3. Deepseek-V3 model

Reference:
1. Deepseek-ai: https://github.com/deepseek-ai/DeepSeek-V3/blob/main/inference/model.py
4. Huggingface: https://huggingface.co/deepseek-ai/DeepSeek-V3-Base/blob/main/modeling_deepseek.py
5. torchtitan/experiment/deepseek-v3
6. torchtitan/experiment/llama4

## TODO
- [ ] Further clean up the DeepseekV3ModelArgs class, remove unused model args
- [ ] Test forward pass w/ torchtitan

### CODE REVIEW

File: torchtitan/models/deepseek-v3/model/model.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        torch.Tensor: Precomputed complex exponential values for positional embeddings.
+    """
+    dim = args.qk_rope_head_dim
+    seqlen = args.max_seq_len
+    beta_fast = args.beta_fast
+    beta_slow = args.beta_slow
+    base = args.rope_theta
+    factor = args.rope_factor
+
+    def find_correction_dim(num_rotations, dim, base, max_seq_len):
```
Comment (fegin) [inline torchtitan/models/deepseek-v3/model/model.py]: nit: typing for args and output.

File: torchtitan/models/deepseek-v3/model/model.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        Returns:
+            float: The correction dimension based on the input parameters.
+        """
+        return (
+            dim
+            * math.log(max_seq_len / (num_rotations * 2 * math.pi))
+            / (2 * math.log(base))
+        )
+
+    def find_correction_range(low_rot, high_rot, dim, base, max_seq_len):
```
Comment (fegin) [inline torchtitan/models/deepseek-v3/model/model.py]: nit: typing for args and output.

File: torchtitan/models/deepseek-v3/model/model.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        self,
+        dim: int,
+        hidden_dim: int,
+    ):
+        super().__init__()
+        self.w1 = nn.Linear(dim, hidden_dim, bias=False)
+        self.w2 = nn.Linear(hidden_dim, dim, bias=False)
+        self.w3 = nn.Linear(dim, hidden_dim, bias=False)
+
+    def forward(self, x):
```
Comment (fegin) [inline torchtitan/models/deepseek-v3/model/model.py]: nit: typing for args and output.

File: torchtitan/models/deepseek-v3/model/model.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+        )
+        self.init_weights()
+
+    def forward(self, tokens: torch.Tensor):
+        """
+        Forward pass for the Transformer model.
+
+        Args:
+            tokens (torch.Tensor): Input tensor of token IDs with shape (batch_size, seq_len).
+            start_pos (int, optional): Starting position in the sequence for rotary embeddings. Defaults to 0.
```
Comment (fegin) [inline torchtitan/models/deepseek-v3/model/model.py]: `start_pos` is not in the arg list?

File: torchtitan/models/deepseek-v3/model/model.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        x: torch.Tensor,
+        freqs_cis: torch.Tensor,
+    ):
+        """
+        Forward pass for the Multi-Head Latent Attention (MLA) Layer.
+
+        Args:
+            x (torch.Tensor): Input tensor of shape (batch_size, seq_len, dim).
+            freqs_cis (torch.Tensor): Precomputed complex exponential values for rotary embeddings.
+            mask (Optional[torch.Tensor]): Mask tensor to exclude certain positions from attention.
```
Comment (fegin) [inline torchtitan/models/deepseek-v3/model/model.py]: mask not in the arg list?

File: torchtitan/models/deepseek-v3/model/model.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+    def forward(self, x):
+        return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+    def init_weights(self, init_std: float):
+        nn.init.trunc_normal_(self.w1.weight, mean=0.0, std=0.02)
+        for linear in (self.w2, self.w3):
+            nn.init.trunc_normal_(linear.weight, mean=0.0, std=init_std)
+
+
+class TransformerBlock(nn.Module):
```
Comment (H-Huang) [inline torchtitan/models/deepseek-v3/model/model.py:262]: nit: it's just called "Block" in the original DS-v3 implementation, not sure how much we care about exact naming

File: torchtitan/models/deepseek-v3/model/model.py [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+    """
+    Transformer block with attention and feed-forward layers.
+    """
+
+    def __init__(self, model_args: DeepseekV3ModelArgs):
+        super().__init__()
+        self.attention = Attention(model_args)
+        self.attention_norm = nn.RMSNorm(model_args.dim, eps=model_args.norm_eps)
+        self.ffn_norm = nn.RMSNorm(model_args.dim, eps=model_args.norm_eps)
+        self.ffn = FeedForward(model_args.dim, model_args.inter_dim)
```
Comment (H-Huang) [inline torchtitan/models/deepseek-v3/model/model.py]: Are we adding the MoE part later? This is only MLP

File: torchtitan/models/deepseek-v3/model/moe.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+from torchtitan.models.llama3 import model
+
+from .args import DeepseekV3ModelArgs
+
+
+# Reference: torchtitan/experiments/llama4/model/
```
Comment (H-Huang) [inline torchtitan/models/deepseek-v3/model/moe.py:14]: just curious, how come we are using the llama4 one instead of the original?

File: torchtitan/models/deepseek-v3/model/model.py (hunk 7 above)
Comment (wwwjn) [reply torchtitan/models/deepseek-v3/model/model.py]: Oh nice catch, this should be MoE

File: torchtitan/models/deepseek-v3/model/moe.py (hunk 8 above)
Comment (wwwjn) [reply torchtitan/models/deepseek-v3/model/moe.py:14]: Mainly because the llama4 version implemented GroupedExperts, which could use `_grouped_mm` to speed up the calculation of experts. The Router and MoE implementation should be almost the same, and I double checked the forward pass and make some changes on llama4 version to align with Deepseek-v3 original version

File: torchtitan/models/deepseek-v3/model/model.py (hunk 6 above)
Comment (wwwjn) [reply torchtitan/models/deepseek-v3/model/model.py:262]: I adopted this name from llama4 as well :)
